google-cloud-firestore/src/main/java/com/google/cloud/firestore/telemetry/MetricsUtil.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.firestore.telemetry;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.InternalApi;
+import com.google.api.gax.tracing.ApiTracerFactory;
+import com.google.cloud.firestore.FirestoreOptions;
+import java.util.List;
+import java.util.logging.Logger;
+import javax.annotation.Nonnull;
+
+/**
+ * A utility interface for metrics collection. Classes that implement this interface may make their
+ * own design choices for how they approach metrics collection. For instance, they may be no-op, or
+ * they may use a particular metrics framework such as OpenTelemetry.
+ */
+@InternalApi
+public interface MetricsUtil {
+  final Logger logger = Logger.getLogger(MetricsUtil.class.getName());
+
+  static final String ENABLE_METRICS_ENV_VAR = "FIRESTORE_ENABLE_METRICS";
+
+  /**
+   * Creates an instance of {@code MetricsUtil}. If the environment variable
+   * `FIRESTORE_ENABLE_METRICS` is set to false or off, an instance of {@link DisabledMetricsUtil}
+   * will be returned. Otherwise, an instance of {@link EnabledMetricsUtil} will be returned.
+   *
+   * @param firestoreOptions The Firestore options that configures client side metrics.
+   * @return An instance of {@code MetricsUtil}.
+   */
+  static MetricsUtil getInstance(@Nonnull FirestoreOptions firestoreOptions) {
+    // TODO(metrics): add debug level logging if possible
+    if (shouldCreateEnabledInstance(firestoreOptions)) {
+      return new EnabledMetricsUtil(firestoreOptions);
+    } else {
+      return new DisabledMetricsUtil();
+    }
+  }
+
+  static boolean shouldCreateEnabledInstance(FirestoreOptions firestoreOptions) {
+    // Start with the value from FirestoreOptions
+    boolean shouldCreateEnabledInstance =
+        firestoreOptions.getOpenTelemetryOptions().isMetricsEnabled();
```

Comment:
as discussed offline, we don't want `isMetricsEnabled()`. This is a large PR already, and ideally we should _disable_ the feature and land it in `main`. IIUC you can disable it by doing this here:

```java
  // TODO(metrics): Look up the enable/disable flag from options before launching
  boolean shouldCreateEnabledInstance = false;
```